compiler/rustc_resolve/src/imports.rs

```diff
@@ -1161,46 +1166,35 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
                     msg,
                 );
             } else {
-                let error_msg = if crate_private_reexport {
-                    format!(
-                        "`{}` is only public within the crate, and cannot be re-exported outside",
-                        ident
-                    )
-                } else {
-                    format!("`{}` is private, and cannot be re-exported", ident)
-                };
-
                 if ns == TypeNS {
-                    let label_msg = if crate_private_reexport {
-                        format!("re-export of crate public `{}`", ident)
-                    } else {
-                        format!("re-export of private `{}`", ident)
-                    };
-
-                    struct_span_err!(self.tcx.sess, import.span, E0365, "{}", error_msg)
-                        .span_label(import.span, label_msg)
-                        .note(format!("consider declaring type or module `{}` with `pub`", ident))
+                    self.tcx
+                        .sess
+                        .create_err(CannotBeReexportedNS {
+                            span: import.span,
+                            ident,
+                            crate_private_reexport,
+                        })
```

Comment:
I think you could avoid using a manual impl and use derive if you had an `if crate_private_reexport` here and called `self.tcx.sess.create_err` on different structs.

Reply:
what case *would* a manual impl be preferred?
I've seen many cases that could be trivially implemented with different structs and am interested when it would be appropriate

Reply:
it's pretty rare at least for me when you have a situation where a manual impl is preferred over derives. You could have places where the existing logic that interacts with diagnostics is complicated with a lot of interacting parts, and that could sometimes prevent you from writing a derived impl.

It's just a preference really, and to me derive looks better. Also it helps with the issue of `crate_private_reexport` looking like a field supplied to the diagnostic message when it isn't.